runtime/compiler/z/codegen/J9TreeEvaluator.cpp

```diff
@@ -2994,36 +3001,40 @@ J9::Z::TreeEvaluator::multianewArrayEvaluator(TR::Node * node, TR::CodeGenerator
-   generateS390BranchInstruction(cg, TR::InstOpCode::BRC, TR::InstOpCode::COND_BRC, node, cFlowRegionEnd);
+   cursor = generateS390BranchInstruction(cg, TR::InstOpCode::BRC, TR::InstOpCode::COND_BRC, node, cFlowRegionEnd);
+   iComment("Init class field and jump to end of ICF.");
 
    // We end up in this region of the ICF if the first dimension is non-zero and the second dimension is zero.
-   generateS390LabelInstruction(cg, TR::InstOpCode::LABEL, node, nonZeroFirstDimLabel);
+   cursor = generateS390LabelInstruction(cg, TR::InstOpCode::LABEL, node, nonZeroFirstDimLabel);
+   iComment("nonZeroFirstDimLabel, 2nd dim length is 0.");
 
    TR::Register *componentClassReg = cg->allocateRegister();
    generateRXInstruction(cg, TR::InstOpCode::LG, node, componentClassReg, generateS390MemoryReference(classReg, offsetof(J9ArrayClass, componentType), cg));
 
-   // Calculate maximum allowable object size in elements and jump to OOL if it's higher than firstDimLenReg.
+   // Calculate maximum allowable object size in elements and jump to OOL if it's lower than firstDimLenReg.
    int32_t elementSize = TR::Compiler->om.sizeofReferenceField();
    uintptr_t maxObjectSize = cg->getMaxObjectSizeGuaranteedNotToOverflow();
    uintptr_t maxObjectSizeInElements = maxObjectSize / elementSize;
-   generateS390CompareAndBranchInstruction(cg, TR::InstOpCode::CL, node, firstDimLenReg, static_cast<int32_t>(maxObjectSizeInElements), TR::InstOpCode::COND_BHR, oolJumpLabel, false);
+   cursor = generateS390CompareAndBranchInstruction(cg, TR::InstOpCode::CL, node, firstDimLenReg, static_cast<int32_t>(maxObjectSizeInElements), TR::InstOpCode::COND_BHR, oolJumpLabel, false);
+   iComment("Jump to oolJumpLabel if 1st dim len > the num of elements a block can fit.");
 
    // Now check to see if we have enough space to do the allocation. If not then jump to OOL code.
    int32_t elementSizeAligned = OMR::align(elementSize, TR::Compiler->om.objectAlignmentInBytes());
    int32_t alignmentCompensation = (elementSize == elementSizeAligned) ? 0 : elementSizeAligned - 1;
    static const uint8_t multiplierToStrideMap[] = {0, 0, 1, 0, 2, 0, 0, 0, 3};
    TR_ASSERT_FATAL(elementSize <= 8, "multianewArrayEvaluator - elementSize cannot be greater than 8!");
-   generateRSInstruction(cg, TR::InstOpCode::SLLK, node, temp1Reg, firstDimLenReg, multiplierToStrideMap[elementSize]);
-   generateRILInstruction(cg, TR::InstOpCode::AFI, node, temp1Reg, static_cast<int32_t>(TR::Compiler->om.contiguousArrayHeaderSizeInBytes()) + alignmentCompensation);
+   generateRSInstruction(cg, TR::InstOpCode::SLLG, node, temp1Reg, firstDimLenReg, multiplierToStrideMap[elementSize]);
+   generateRILInstruction(cg, TR::InstOpCode::AGFI, node, temp1Reg, static_cast<int32_t>(TR::Compiler->om.contiguousArrayHeaderSizeInBytes()) + alignmentCompensation);
```

Comment:
Just to confirm, this is the actual fix right?

Reply:
Yup, that is the actual fix.

Reply:
Leaving this review unresolved for future in case someone looks at this PR it is easy to spot to fix itself.

Reply:
Thank you for this!